pkg/expr/translate/translate.go

```diff
@@ -0,0 +1,343 @@
+			queryObj["datasourceUid"] = getDsInfo.Result.Uid
+			queryObj["refId"] = refID
+
+			if _, found := queryObj["maxDataPoints"]; !found {
+				queryObj["maxDataPoints"] = 100
+			}
+			if _, found := queryObj["intervalMs"]; !found {
+				queryObj["intervalMs"] = 1000
+			}
+
+			encodedObj, err := json.Marshal(queryObj)
+			if err != nil {
+				return nil, err
+			}
+
+			rawFrom := newRefIDsToTimeRanges[refID][0]
+			rawTo := newRefIDsToTimeRanges[refID][1]
+
+			rTR, err := getRelativeDuration(rawFrom, rawTo)
+			if err != nil {
+				return nil, err
+			}
+
+			alertQuery := ngmodels.AlertQuery{
+				RefID:             refID,
+				Model:             encodedObj,
+				RelativeTimeRange: *rTR,
+				DatasourceUID:     getDsInfo.Uid,
+			}
+			ngCond.QueriesAndExpressions = append(ngCond.QueriesAndExpressions, alertQuery)
+		}
+	}
+
+	// build the new classic condition pointing our new equivalent queries
+	conditions := make([]classic.ClassicConditionJSON, len(dc.Conditions))
+	for i, cond := range dc.Conditions {
+		newCond := classic.ClassicConditionJSON{}
+		newCond.Evaluator = classic.ConditionEvalJSON{
+			Type:   cond.Evaluator.Type,
+			Params: cond.Evaluator.Params,
+		}
+		newCond.Operator.Type = cond.Operator.Type
+		newCond.Query.Params = append(newCond.Query.Params, condIdxToNewRefID[i])
+		newCond.Reducer.Type = cond.Reducer.Type
+
+		conditions[i] = newCond
+	}
+
+	ccRefID, err := getNewRefID(newRefIDstoCondIdx) // get refID for the classic condition
+	if err != nil {
+		return nil, err
+	}
+	ngCond.RefID = ccRefID // set the alert condition to point to the classic condition
+	ngCond.OrgID = orgID
+
+	exprModel := struct {
+		Type       string                         `json:"type"`
+		RefID      string                         `json:"refId"`
+		Datasource string                         `json:"datasource"`
+		Conditions []classic.ClassicConditionJSON `json:"conditions"`
+	}{
+		"classic_conditions",
+		ccRefID,
+		"__expr__",
+		conditions,
+	}
+
+	exprModelJSON, err := json.Marshal(&exprModel)
+	if err != nil {
+		return nil, err
+	}
+
+	ccAlertQuery := ngmodels.AlertQuery{
+		RefID: ccRefID,
+		Model: exprModelJSON,
+	}
+
+	ngCond.QueriesAndExpressions = append(ngCond.QueriesAndExpressions, ccAlertQuery)
+
+	for i := range ngCond.QueriesAndExpressions {
+		err := ngCond.QueriesAndExpressions[i].PreSave()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	sort.Slice(ngCond.QueriesAndExpressions, func(i, j int) bool {
+		return ngCond.QueriesAndExpressions[i].RefID < ngCond.QueriesAndExpressions[j].RefID
+	})
+
+	return ngCond, nil
+}
+
+const alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+// getNewRefID finds first capital letter in the alphabet not in use
+// to use for a new RefID.
+func getNewRefID(refIDs map[string][]int) (string, error) {
+	for _, r := range alpha {
+		sR := string(r)
+		if _, ok := refIDs[sR]; ok {
+			continue
+		}
+		return sR, nil
+	}
+	return "", fmt.Errorf("ran out of letters when creating expression")
```

Comment:
Hmm.. I can check if there is a limit. If there is a length limit, now that you point it out this wouldn't be quite correct anyways I think since they may be ways that the translation exceeds the limit (e.g. someone has recently discovered rename refId).

I may keep it for now and add a comment. But have to see if there is a length limit, and if there isn't, will add more somethings like you say.

Reply:
went with comment for now, can fix in future